plugins/inputs/zfs/zfs_freebsd.go

```diff
@@ -0,0 +1,142 @@
+package zfs
+
+import (
+	"bytes"
+	"fmt"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+// $ zpool list -Hp
+// freenas-boot    30601641984     2022177280      28579464704     -       -       6       1.00x   ONLINE  -
+// red1    8933531975680   1126164848640   7807367127040   -       8%      12      1.83x   ONLINE  /mnt
+// temp1   2989297238016   1626309320704   1362987917312   -       38%     54      1.28x   ONLINE  /mnt
+// temp2   2989297238016   626958278656    2362338959360   -       12%     20      1.00x   ONLINE  /mnt
+
+func gatherPoolStats(poolStats bool, acc telegraf.Accumulator) (string, error) {
+
+	lines, err := run("zpool", []string{"list", "-Hp"}...)
+	if err != nil {
+		return "", err
+	}
+
+	pools := []string{}
+	for _, line := range lines {
+		col := strings.Split(line, "\t")
+
+		pools = append(pools, col[0])
+	}
+
+	if poolStats {
+		for _, line := range lines {
+			col := strings.Split(line, "\t")
+			tags := map[string]string{"pool": col[0], "health": col[8]}
+			fields := map[string]interface{}{}
+
+			size, err := strconv.ParseInt(col[1], 10, 64)
+			if err != nil {
+				return "", fmt.Errorf("Error parsing size: %s", err)
+			}
+			fields["size"] = size
+
+			alloc, err := strconv.ParseInt(col[2], 10, 64)
+			if err != nil {
+				return "", fmt.Errorf("Error parsing alloc: %s", err)
+			}
+			fields["alloc"] = alloc
+
+			free, err := strconv.ParseInt(col[3], 10, 64)
+			if err != nil {
+				return "", fmt.Errorf("Error parsing free: %s", err)
+			}
+			fields["free"] = free
+
+			frag, err := strconv.ParseInt(strings.TrimSuffix(col[5], "%"), 10, 0)
+			if err == nil {
+				// This might be - for RO devs
+				fields["frag"] = frag
+			}
+
+			capval, err := strconv.ParseInt(col[6], 10, 0)
+			if err != nil {
+				return "", fmt.Errorf("Error parsing cap: %s", err)
+			}
+			fields["cap"] = capval
+
+			dedup, err := strconv.ParseFloat(strings.TrimSuffix(col[7], "x"), 32)
+			if err != nil {
+				return "", fmt.Errorf("Error parsing dedup: %s", err)
+			}
+			fields["dedup"] = dedup
+
+			acc.AddFields("zfs_pool", fields, tags)
+		}
+	}
+
+	return strings.Join(pools, "::"), nil
+}
+
+func (z *Zfs) Gather(acc telegraf.Accumulator) error {
+	kstatMetrics := z.KstatMetrics
+	if len(kstatMetrics) == 0 {
+		kstatMetrics = []string{"arcstats", "zfetchstats", "vdev_cache_stats"}
+	}
+
+	// TODO handle this better.
+	kstatPath := z.KstatPath
+	if len(kstatPath) > 0 {
+		return fmt.Errorf("Configuration 'kstatPath' is not supported on FreeBSD.")
+	}
```

Comment:
@sparrc 
> 2. Can you explain more or give an example? How are the configs different between freebsd and linux?

`kstatPath` is not used on FreeBSD since it don't have proc fs of the type Linux have, instead kstats are read with `sysctl`.

The current code will return a error if `kstatPath` is defined but we could also just warn or ignore. Alternatively the plugin could use different config structs and `sampleConfig`.